### PR TITLE
add more ctc-loss related summaries

### PR DESCRIPTION
**Pull request by yqwangustc**
Add more detailed loss to easy debugging

### CODE REVIEW

File: axlearn/audio/asr_decoder.py [hunk 1]
```diff
@@ -196,19 +196,30 @@ def _compute_target_paddings(
...
+        # pytype: disable=attribute-error
+        ret_dict = {
+            "input_stats/average_target_length": WeightedScalar(
+                jnp.mean(target_lengths), batch_size
+            ),
+            "input_stats/average_source_length": WeightedScalar(
+                jnp.mean(source_lengths), batch_size
+            ),
+            "input_stats/frame_packing_effiency": WeightedScalar(
+                jnp.sum(source_lengths) / input_batch["paddings"].size, input_batch["paddings"].size
```
Comment (markblee) [inline axlearn/audio/asr_decoder.py:218]: Guard against division by 0 here and below?

File: axlearn/audio/asr_decoder.py [hunk 2]
```diff
@@ -196,19 +196,30 @@ def _compute_target_paddings(
...
+            ),
+        }
+        # pytype: enable=attribute-error
+        return ret_dict
```
Comment (markblee) [inline axlearn/audio/asr_decoder.py:222]: Is there a specific reason to prefer returning the summaries instead of just adding them here?

File: axlearn/audio/asr_decoder.py [hunk 3]
```diff
@@ -196,19 +196,30 @@ def _compute_target_paddings(
         target_labels: Tensor = input_batch["target_labels"]
         # Infer target_paddings from out-of-range labels.
         target_paddings = jnp.logical_or(cfg.vocab_size <= target_labels, target_labels < 0)
+        return target_paddings
 
+    def _input_stats_summary(
```
Comment (markblee) [inline axlearn/audio/asr_decoder.py:201]:
```suggestion
    def _input_stats_summaries(
```
or `def _add_input_stats_summaries` if we decide to inline the add, which may be more similar to other callsites in the repo.

File: axlearn/audio/asr_decoder.py [hunk 4]
```diff
@@ -257,6 +268,64 @@ def predict(self, input_batch: Nested[Tensor]) -> Tensor:
...
+        target_paddings: Tensor,
+    ) -> Dict[str, Union[WeightedScalar, Tensor]]:
+        valid_frame_mask = (1.0 - paddings) * per_example_weight[:, None]
+        valid_label_mask = (1.0 - target_paddings) * per_example_weight[:, None]
+
+        num_valid_frames = jnp.sum(valid_frame_mask)
+        num_valid_labels = jnp.sum(valid_label_mask)
+        per_frame_loss = total_ctc_loss / num_valid_frames
+        per_label_loss = total_ctc_loss / num_valid_labels
+        batch_size = per_example_weight.shape[0]
```
Comment (markblee) [inline axlearn/audio/asr_decoder.py:311]: Same here?

File: axlearn/audio/asr_decoder.py (hunk 1 above)
Comment (yqwangustc) [reply axlearn/audio/asr_decoder.py:218]: I think this before: the denominator `input_batch['paddings'].size` is directly pulled from the input. If it is zero, many other places will become problematic well before we hit here (e.g., various normalization layer). Nevertheless, I have added safeguard for them.

File: axlearn/audio/asr_decoder.py (hunk 4 above)
Comment (yqwangustc) [reply axlearn/audio/asr_decoder.py:311]: Done.

File: axlearn/audio/asr_decoder.py (hunk 1 above)
Comment (markblee) [reply axlearn/audio/asr_decoder.py:218]: Thanks! I don't think we should make assumptions about the other components.

File: axlearn/audio/asr_decoder.py (hunk 3 above)
Comment (markblee) [reply axlearn/audio/asr_decoder.py:201]: Have the changes been pushed?

Comment (yqwangustc) [reply axlearn/audio/asr_decoder.py]: sorry. was confused and pushed to the other repo. Just pushed the change.

File: axlearn/audio/asr_decoder.py [hunk 5]
```diff
@@ -257,6 +269,65 @@ def predict(self, input_batch: Nested[Tensor]) -> Tensor:
...
+        paddings: Tensor,
+        target_paddings: Tensor,
+    ) -> Dict[str, Union[WeightedScalar, Tensor]]:
+        valid_frame_mask = (1.0 - paddings) * per_example_weight[:, None]
+        valid_label_mask = (1.0 - target_paddings) * per_example_weight[:, None]
+
+        num_valid_frames = jnp.sum(valid_frame_mask)
+        num_valid_labels = jnp.sum(valid_label_mask)
+        per_frame_loss = total_ctc_loss / num_valid_frames
+        per_label_loss = total_ctc_loss / num_valid_labels
```
Comment (markblee) [inline axlearn/audio/asr_decoder.py:312]: Here too?

**Review by markblee — APPROVED** [review]
Thanks, a few nits.

File: axlearn/audio/asr_decoder.py [hunk 6]
```diff
@@ -257,6 +269,65 @@ def predict(self, input_batch: Nested[Tensor]) -> Tensor:
...
+    def _input_stats_summaries(
+        self, input_batch: Nested[Tensor], per_example_weight: Tensor
+    ) -> Dict[str, Union[WeightedScalar, Tensor]]:
+        paddings = input_batch["paddings"]
+        target_paddings = self._compute_target_paddings(input_batch)
+        valid_frame_mask = (1.0 - paddings) * per_example_weight[:, None]
+        valid_label_mask = (1.0 - target_paddings) * per_example_weight[:, None]
+        num_valid_frames = jnp.sum(valid_frame_mask)
+        num_valid_labels = jnp.sum(valid_label_mask)
+        num_valid_examples = jnp.maximum(per_example_weight.sum(), 1.0)
```
[the remark this answers is not in the document]
Comment (yqwangustc) [reply axlearn/audio/asr_decoder.py:281]: Thanks for the suggestion.

File: axlearn/audio/asr_decoder.py (hunk 5 above)
Comment (yqwangustc) [reply axlearn/audio/asr_decoder.py:312]: Done.

**Review by markblee — APPROVED** [review]
